byteparse/shared/src/main/scala/byteparse/classparse/ClassParser.scala

```diff
@@ -0,0 +1,456 @@
+    BasicElemInfo(DoubleElem(wrapByteBuffer(bs).getDouble)))
+
+  val constantNameAndTypeInfo =
+    P( BS(12) ~
+       AnyWordI /*name_index*/ ~
+       AnyWordI /*descriptor_index*/
+    ).map(NameAndTypeInfo.tupled)
+
+  val constantUtf8Info =
+    P( BS(1) ~
+       AnyWordI.flatMap(l =>
+         P( AnyByte.rep(exactly=l).! ))
+    ).map(Utf8Info)
+
+  val constantMethodHandleInfo =
+    P( BS(15) ~
+       AnyByte.! /*reference_kind*/ ~
+       AnyWordI /*reference_index*/
+    ).map {
+      case (refKind, refIdx) => MethodHandleInfo(refKind(0).toInt, refIdx)
+    }
+
+  val constantMethodTypeInfo =
+    P( BS(16) ~ AnyWordI /*descriptor_index*/ )
+      .map(MethodTypeInfo)
+
+  val constantInvokeDynamicInfo =
+    P( BS(18) ~
+       AnyWordI /*bootstrap_method_attr_index*/ ~
+       AnyWordI /*name_and_type_index*/
+     ).map(InvokeDynamicInfo.tupled)
+
+  val singleConstantPoolItem = P( constantClassInfo      | constantFieldRefInfo |
+                                  constantMethodRefInfo  | constantInterfaceMethodRefInfo |
+                                  constantStringInfo     | constantIntInfo |
+                                  constantFloatInfo      | constantNameAndTypeInfo |
+                                  constantUtf8Info       | constantMethodHandleInfo |
+                                  constantMethodTypeInfo | constantInvokeDynamicInfo )
+
+  val doubleConstantPoolItem = P( constantDoubleInfo | constantLongInfo)
+
+  def constantPool(count: Int): P[Seq[PoolInfo]] =
+    if (count == 0) {
+      Pass.map(_ => ArrayBuffer[PoolInfo]())
+    } else {
+        P( singleConstantPoolItem.flatMap(item => constantPool(count - 1).map(_ :+ item)) |
+           doubleConstantPoolItem.flatMap(item => constantPool(count - 2).map(_ :+ NopInfo() :+ item )) )
+    }
+
+  val attributeInfo =
+    P( AnyWordI /*attribute_name_index*/ ~
+       AnyDwordI /*attribute_length*/.flatMap(l =>
+         P( AnyByte.rep(exactly=l).! ))
+     ).map(AttributeInfo.tupled)
+
+
+  val fieldInfo =
+    P( AnyWord.! /*access_flags*/ ~
+       AnyWordI /*name_index*/ ~
+       AnyWordI /*descriptor_index*/ ~
+       AnyWordI /*attributes_count*/.flatMap(l =>
+         P( attributeInfo.rep(exactly=l) ))
+     ).map(FieldInfo.tupled)
+
+  val methodInfo =
+    P( AnyWord.! /*access_flags*/ ~
+       AnyWordI /*name_index*/ ~
+       AnyWordI /*descriptor_index*/ ~
+       AnyWordI /*attributes_count*/.flatMap(l =>
+         P( attributeInfo.rep(exactly=l) ))
+     ).map(MethodInfo.tupled)
+
+  val classFile =
+    P( BS(0xCA, 0xFE, 0xBA, 0xBE) ~
+       AnyWordI /*minor_version*/ ~
+       AnyWordI /*major_version*/ ~
+       AnyWordI /*constant_pool_count*/.flatMap(l =>
+         constantPool(l - 1).map(_.reverse)) ~
+       AnyWord.! /*access_flags*/ ~
+       AnyWordI /*this_class*/ ~
+       AnyWordI /*super_class*/ ~
+       AnyWordI /*interfaces_count*/.flatMap(l =>
+         P( AnyWordI.rep(exactly=l) )) ~
```

Comment:
We probably don't need to wrap these inner-parsers in `P(...)`s. `P(...)` doesn't actually change the type of things; the main purpose of `P(...)` is to 

- provide a human-readable marker in the error-stack, and 
- to allow recursion. 

The first one doesn't really apply here (I think it'll just pick up the same `classFile` name as the outer `P`, which is not very helpful) and neither does the second (since there's no recursion here) so I'd just drop it